tools/tsTreeshaker.ts

```diff
@@ -0,0 +1,258 @@
+
+    this.graph = new DependencyGraph([...declarationRoots, ...jsRoots]);
+  }
+  addNode(filePath: string, requestedModule: string, extension: '.d.ts' | '.js') {
+    if (this.graph === undefined) throw new Error('graph is undefined');
+
+    requestedModule = this.getAbsolutePath(filePath, requestedModule);
+    requestedModule = requestedModule + this.getFileExtension(requestedModule);
+    requestedModule = this.replaceExtension(requestedModule, extension);
+
+    filePath = this.replaceExtension(filePath, extension);
+
+    requestedModule = this.convertToDestinationPath(requestedModule);
+    requestedModule = this.normalizePath(requestedModule);
+    filePath = this.convertToDestinationPath(filePath);
+    filePath = this.normalizePath(filePath);
+
+    this.graph.addDependency(filePath, requestedModule);
+  }
+  getIncludedDependencies() {
+    if (this.graph === undefined) throw new Error('graph is undefined');
+    return this.graph.getIncludedDependencies();
+  }
+}
+
+class TsUtils {
+  static isRequireCall(
+    node: ts.Node,
+    checkArgumentIsStringLiteralLike: boolean
+  ): node is ts.CallExpression {
+    if (!ts.isCallExpression(node)) {
+      return false;
+    }
+    const { expression, arguments: args } = node;
+    if (!ts.isIdentifier(expression) || expression.escapedText !== 'require') {
+      return false;
+    }
+    if (args.length !== 1) {
+      return false;
+    }
+    const [arg] = args;
+
+    return !checkArgumentIsStringLiteralLike || ts.isStringLiteralLike(arg);
+  }
+  static isImportCall(node: ts.Node): node is ts.CallExpression {
+    return (
+      ts.isCallExpression(node) && node.expression.kind === ts.SyntaxKind.ImportKeyword
+    );
+  }
+  static chainBundle<T extends ts.SourceFile | ts.Bundle>(
+    transformSourceFile: (x: ts.SourceFile) => ts.SourceFile | undefined
+  ): (x: T) => T {
+    function transformBundle(node: ts.Bundle) {
+      return factory.createBundle(
+        node.sourceFiles.map(transformSourceFile).filter((x): x is ts.SourceFile => !!x),
+        node.prepends
+      );
+    }
+
+    return function transformSourceFileOrBundle(node: T) {
+      return ts.isSourceFile(node)
+        ? (transformSourceFile(node) as T)
+        : (transformBundle(node as ts.Bundle) as T);
+    };
+  }
+}
+
+export type transformerConfig = {
+  solver: DependencySolver;
+};
+
+export default function transformer(config: transformerConfig) {
+  return {
+    after: [(context: ts.TransformationContext) => {
+      config.solver.setCompilerOptions(context.getCompilerOptions());
+      return new TransformerBuilder(config.solver, '.js').make<ts.SourceFile>(context);
+    }],
+    afterDeclarations: [(context: ts.TransformationContext) => {
+      config.solver.setCompilerOptions(context.getCompilerOptions());
+      return new TransformerBuilder(config.solver, '.d.ts').make(context);
+    }]
+  };
+}
+
+class TransformerBuilder {
```

Comment:
Why use class?

Reply:
Because the superposition of so many closure made it hard to read the code